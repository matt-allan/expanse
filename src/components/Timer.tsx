import React, { useState } from 'react';
import { Grommet, Box, Button, Clock, Heading, Meter } from 'grommet';
import { Play, Pause, Resume, Refresh } from "grommet-icons";

const DURATION_REGEXP = /P([0-9]+)H([0-9]+)M([0-9]+)S/;

const SECONDS_PER_SESSION = 25 * 60;

const parseDuration = (duration: string): number => {
    let match = DURATION_REGEXP.exec(duration);

    if (!match) {
      throw `Cannot parse '${duration}'`;
    }

    const hours = parseFloat(match[1]);
    const minutes = parseFloat(match[2]) || 0;
    const seconds = parseFloat(match[3]) || 0;

    return (hours * 60 * 60) + (minutes * 60) + seconds;
}

export const Timer = () => {
  const [paused, setPaused] = useState(true);

  const [secondsLeft, setSecondsLeft] = useState(SECONDS_PER_SESSION);

  return (
    <React.Fragment>
      <Box align="center" justify="center" pad={{ top: 'large', bottom: 'small'}}>
        <Meter
          type="circle"
          background="light-2"
          max={SECONDS_PER_SESSION}
          values={[{ value: secondsLeft }]}
        />
      </Box>
      <Box align="center" justify="center" pad="small">
        <Clock
          key={secondsLeft == SECONDS_PER_SESSION ? 'restarted' : 'running'}
          type="digital"
          time={`PT0H${SECONDS_PER_SESSION / 60}M0S`}
          run={paused ? false : 'backward'}
          onChange={(time: any) => setSecondsLeft(parseDuration(time))}
          alignSelf="center"
          size="xlarge"
          margin="xlarge"
          a11yTitle="countdown timer until next break"
        />
      </Box>
      <Box direction="row" justify="center" gap="medium" pad={{top: 'small', bottom: 'large'}}>
        <Button plain={true} icon={<Refresh />} onClick={() => setSecondsLeft(SECONDS_PER_SESSION)} />
        <Button plain={true} icon={paused ? <Play /> : <Pause />} onClick={() => setPaused(!paused)} />
      </Box>
    </React.Fragment>
  );
}