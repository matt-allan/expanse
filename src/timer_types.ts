export enum Status {
  Started = "started",
  Stopped = "stopped",
  Ended = "ended",
}

export const statuses = Object.values(Status);

export enum Event {
  Started = "started",
  Stopped = "stopped",
  Restarted = "restarted",
  Ended = "ended",
  Tick = "tick",
}

export const events = Object.values(Event);
