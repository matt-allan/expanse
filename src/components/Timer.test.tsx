import React from "react";
import { shallow } from "enzyme";

import { Timer } from "./Timer";
import { Status } from "./../timer_types";

jest.useFakeTimers();

it("renders", () => {
  const component = shallow(
    <Timer
      seconds={5 * 60}
      remaining={5 * 60}
      status={Status.Stopped}
      onStart={jest.fn()}
      onStop={jest.fn()}
      onRestart={jest.fn()}
    />
  );

  expect(component).toMatchSnapshot();
});
