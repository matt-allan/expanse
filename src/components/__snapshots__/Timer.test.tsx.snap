// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Fragment>
  <Box
    align="center"
    justify="center"
    pad={
      Object {
        "bottom": "small",
        "top": "large",
      }
    }
  >
    <Meter
      background="light-2"
      max={300}
      type="circle"
      values={
        Array [
          Object {
            "value": 300,
          },
        ]
      }
    />
  </Box>
  <Box
    align="center"
    justify="center"
    pad="small"
  >
    <Clock
      a11yTitle="countdown timer until next break"
      alignSelf="center"
      margin="xlarge"
      run={false}
      size="xlarge"
      time="PT0H5M0S"
      type="digital"
    />
  </Box>
  <Box
    direction="row"
    gap="medium"
    justify="center"
    pad={
      Object {
        "bottom": "large",
        "top": "small",
      }
    }
  >
    <Button
      icon={<Refresh />}
      onClick={[MockFunction]}
      plain={true}
    />
    <Button
      icon={<Play />}
      onClick={[MockFunction]}
      plain={true}
    />
  </Box>
</Fragment>
`;
